File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.max_so_far = float('-inf')
        def dfs(curr):
            if not curr:
                return 0
            left_max = max(dfs(curr.left), 0)
            right_max = max(dfs(curr.right), 0)

            curr_path = curr.val + left_max + right_max
            self.max_so_far = max(curr_path, self.max_so_far)
            return curr.val + max(left_max, right_max)
        
        dfs(root)

        return self.max_so_far

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private int count = 0;
    public int goodNodes(TreeNode root) {
        dfs(root, root.val);
        return count;
    }

    private void dfs(TreeNode cur, int maxSoFar) {
        if (cur == null) return;
        int newMax = Math.max(cur.val, maxSoFar);
        if (cur.val >= newMax) count++;
        dfs(cur.left, newMax);
        dfs(cur.right, newMax);
    }
}
